src/black/linegen.py

```diff
@@ -866,38 +884,62 @@ def normalize_invisible_parens(
             elif not (isinstance(child, Leaf) and is_multiline_string(child)):
                 wrap_in_parentheses(node, child, visible=False)
 
-        check_lpar = isinstance(child, Leaf) and child.value in parens_after
+        comma_check = child.type == token.COMMA if preview else False
+
+        check_lpar = isinstance(child, Leaf) and (
+            child.value in parens_after or comma_check
+        )
+
+
+def remove_with_parens(node: Node, parent: Node) -> None:
+    """Recursively hide optional parens in `with` statements."""
+    if node.type == syms.atom:
+        if maybe_make_parens_invisible_in_atom(
+            node,
+            parent=parent,
+            remove_brackets_around_comma=True,
+        ):
+            wrap_in_parentheses(parent, node, visible=False)
+        if isinstance(node.children[1], Node):
+            remove_with_parens(node.children[1], node)
+    elif node.type == syms.testlist_gexp:
+        for child in node.children:
+            if isinstance(child, Node):
+                remove_with_parens(child, node)
+    elif node.type == syms.asexpr_test and not any(
+        leaf.type == token.COLONEQUAL for leaf in node.leaves()
+    ):
+        if maybe_make_parens_invisible_in_atom(
+            node.children[0],
+            parent=parent,
+            remove_brackets_around_comma=True,
+        ):
+            wrap_in_parentheses(parent, node.children[0], visible=False)
```

Comment:
`with` is slightly more complicated to achieve in one pass so I've split this into it's own function.

Basically the different variations of bracketed with statements give pretty different parse trees so we need to recursively solve this.

`with (open("test.txt")) as f: ...`: this is an `asexpr_test`
`with (open("test.txt") as f): ...`: this is an `atom` containing an `asexpr_test`
`with (open("test.txt")) as f, (open("test.txt")) as f: ...`: this is `asexpr_test`, `COMMA`, `asexpr_test`
`with (open("test.txt") as f, open("test.txt")) as f: ...`: this is an `atom` containing a `testlist_gexp` which then contains multiple `asexpr_test`s